include/dali/core/mm/async_pool.h

```diff
@@ -71,6 +71,19 @@ class async_pool_resource : public async_memory_resource<Kind> {
           (e.is_drv_api() && e.drv_error() != CUDA_ERROR_DEINITIALIZED))
         std::terminate();
     }
+    for (auto &kv : stream_free_) {
+      PerStreamFreeBlocks &free_blocks = kv.second;
+
+      // find_first_ready doesn't throw on shutdown - and we want to terminate on other errors
+      auto *first_free = find_first_ready(free_blocks);
+      assert(first_free == free_blocks.free_list.head);
```

Comment:
What we get inside `first_free` when there were no allocations?

Reply:
Why you get `first_free` here? Just for an assert?

Reply:
1. nullptr
2. Yes, just for that - moved to inside of the assert.